### ISSUE TITLE
Dedicated server loads log4j2.xml from vanilla jar

### ISSUE THREAD

**Issue by IzzelAliz** [MinecraftForge/MinecraftForge#8274]
**Minecraft Version:** 1.17/1.18

**Forge Version:** 38.0.17 whatever

**Logs:** -

**Steps to Reproduce:**
 1. Install forge server
 2. and just start

**Description of issue:**

Dedicated server loads log4j2.xml from vanilla jar, results to missing of debug.log, missing of console color.

**Comment by sciwhiz12** [comment]
Confirmed on 1.18.1-39.0.5 by observing the absence of the `debug.log` and debugging level logging, and by adding the system property `log4j2.debug`[^1] to the command line (using `-Dlog4j2.debug`) and reading the console output to find the relevant initialization lines from Log4J2:
```
TRACE StatusLogger Trying to find [log4j2.jsn] using cpw.mods.cl.ModuleClassLoader@6c49835d class loader.
TRACE StatusLogger Trying to find [log4j2.jsn] using ClassLoader.getSystemResource().
TRACE StatusLogger Trying to find [log4j2.xml] using context class loader cpw.mods.cl.ModuleClassLoader@6c49835d.
DEBUG StatusLogger Not in a ServletContext environment, thus not loading WebLookup plugin.
DEBUG StatusLogger Closing ChannelInputStream sun.nio.ch.ChannelInputStream@53976f5c
DEBUG StatusLogger Apache Log4j Core 2.15.0 initializing configuration XmlConfiguration[location=union:/D:/Dev/Forge/Server/libraries/net/minecraft/server/1.18.1-20211210.034407/server-1.18.1-20211210.034407-extra.jar%2324!/log4j2.xml]
```

`D:/Dev/Forge/Server/libraries/net/minecraft/server/1.18.1-20211210.034407/server-1.18.1-20211210.034407-extra.jar!/log4j2.xml` is the log4j2 configuration file from the vanilla server. The log4j2 configuration file for Forge would be located, for example, at `D:/Dev/Forge/Server/libraries/net/minecraftforge/fmlloader/1.18.1-39.0.5/fmlloader-1.18.1-39.0.5.jar!/log4j2.xml`.

[^1]: [Apache Log4j 2 - Configuration - System Properties](https://logging.apache.org/log4j/2.x/manual/configuration.html#SystemProperties): `log4j2.debug` - "Log4j2 will print all internal logging to the console if system property log4j2.debug is either defined empty or its value equals to true (ignoring case)."

**Comment by sciwhiz12** [comment]
I've tracked down the bug to likely caused by [this use of `Collectors#toMap` in the constructor of `ModuleClassLoader`][mcl_ctor] in securejarhandler, for the `resolvedRoots` field. 

My tests show that, when stored in a `HashMap`, the string key `server` is before `fmlloader` in the (unguaranteed) ordering for `HashMap`. The module `server` contains the vanilla dedicated server JAR (which has the vanilla log4j conf. file), and the module `fmlloader` contains the FML loader JAR (which has the Forge-provided log4j2 conf. file).

Because `log4j.xml` does not exist in a package of a module, [the `ModuleClassLoader#findResourceList`][mcl_findResourceList] code falls back to the `resolvedRoots` field and iterates over this map. Because `server` is before `fmlloader`, the `log4j.xml` from the module for the vanilla dedicated server JAR will be ahead of the `log4j.xml` from the module from FML loader, and therefore the log4j conf. file from the vanilla dediserver is what is used by Log4j.

A possible (as of yet untested) solution would be amending that call to `Collectors#toMap` for the `resolvedRoots` field to an overload to customize the map returned to be a `LinkedHashMap` instead.

[mcl_ctor]: https://github.com/MinecraftForge/securejarhandler/blob/f81a31abecf63d567a9741f11a7b227dbb1a45fb/src/main/java/cpw/mods/cl/ModuleClassLoader.java#L36-L39
[mcl_findResourceList]: https://github.com/MinecraftForge/securejarhandler/blob/f81a31abecf63d567a9741f11a7b227dbb1a45fb/src/main/java/cpw/mods/cl/ModuleClassLoader.java#L196-L201

**Comment by sciwhiz12** [comment]
Confirmed fixed on `1.18.2-40.0.2`. The used Log4J XML configuration (as logged through `log4j2.debug`) is now the configuration from the `fmlloader` JAR, rather than the vanilla dedicated server, which means the `debug.log` now generates as expected.

However, it bears noting that coloring and debug level logging in the dedicated server's console seems to still be absent (at least in my testing. As a bit of speculation, I presume the former is likely due to our outdated JLine and TerminalConsoleAppender dependencies (3.12.+ and 1.2.0, vs. the latest 3.21.0 and 1.3.0).
